src/goto-checker/multi_path_symex_checker.cpp

```diff
@@ -0,0 +1,248 @@
+
+#include <chrono>
+
+#include "bmc_util.h"
+#include "counterexample_beautification.h"
+
+multi_path_symex_checkert::multi_path_symex_checkert(
+  const optionst &options,
+  ui_message_handlert &ui_message_handler,
+  abstract_goto_modelt &goto_model)
+  : multi_path_symex_only_checkert(options, ui_message_handler, goto_model),
+    equation_generated(false)
+{
+  solver_factoryt solvers(
+    options,
+    ns,
+    ui_message_handler,
+    ui_message_handler.get_ui() == ui_message_handlert::uit::XML_UI);
+  solver = solvers.get_solver();
+  solver->prop_conv().set_message_handler(ui_message_handler);
+}
+
+incremental_goto_checkert::resultt multi_path_symex_checkert::
+operator()(propertiest &properties)
+{
+  resultt result(resultt::progresst::DONE);
+
+  // When the equation has been generated, we know all the properties.
+  // Have we got anything to check? Otherwise we return DONE.
+  if(equation_generated && !has_properties_to_check(properties))
+    return result;
+
+  prop_convt &prop_conv = solver->prop_conv();
+  std::chrono::duration<double> solver_runtime(0);
+
+  // we haven't got an equation yet
+  if(!equation_generated)
+  {
+    perform_symex();
+    output_coverage_report();
+    // This might add new properties such as unwinding assertions, for instance.
+    update_properties_status_from_symex_target_equation(
+      properties, result.updated_properties, equation);
+    // Since we will not symex any further we can decide the status
+    // of all properties that do not occur in the equation now.
+    // The current behavior is PASS.
+    update_status_of_not_checked_properties(
+      properties, result.updated_properties);
+
+    // Have we got anything to check? Otherwise we return DONE.
+    if(!has_properties_to_check(properties))
+      return result;
+
+    log.status() << "Passing problem to " << prop_conv.decision_procedure_text()
+                 << messaget::eom;
+
+    const auto solver_start = std::chrono::steady_clock::now();
+
+    convert_symex_target_equation(equation, prop_conv, ui_message_handler);
+    update_properties_goals_from_symex_target_equation(properties);
+    convert_goals();
+    freeze_goal_variables();
+
+    const auto solver_stop = std::chrono::steady_clock::now();
+    solver_runtime += std::chrono::duration<double>(solver_stop - solver_start);
+
+    log.status() << "Running " << prop_conv.decision_procedure_text()
+                 << messaget::eom;
+    equation_generated = true;
+  }
+
+  const auto solver_start = std::chrono::steady_clock::now();
+
+  add_constraint_from_goals(properties);
+
+  decision_proceduret::resultt dec_result = prop_conv.dec_solve();
+
+  update_properties_status_from_goals(
+    properties, result.updated_properties, dec_result);
+
+  const auto solver_stop = std::chrono::steady_clock::now();
+  solver_runtime += std::chrono::duration<double>(solver_stop - solver_start);
+  log.status() << "Runtime decision procedure: " << solver_runtime.count()
```

Comment:
While we're here: "Decision procedure runtime:"

Reply:
https://github.com/diffblue/cbmc/issues/4013

Please drop further suggestions there, @smowton.